distml/strategy/ps_strategy.py

```diff
@@ -0,0 +1,702 @@
+        # All sync with worker 0
+        init_weights_id = self.worker_group.get_named_parameters(cpu=True)
+
+        self._round_robin_sharding()
+
+        # set assignments to every worker
+        self.worker_group.set_assignments(self.assignments)
+
+        # all workers get synced
+        for i, worker in enumerate(self.worker_group.actors):
+            if i != 0:
+                ray.get([worker.set_parameters.remote(init_weights_id)])
+
+        # now spawn parameter server actors
+        shard_ids = self.worker_group.split_parameters(self.assignments)
+
+        # TODO(HUI): use scatter to send parameters
+        for server_idx, server in enumerate(self.server_group.actors):
+            this_shard_ref = self.worker_group.actors[0].index_shard.remote(
+                shard_ids, server_idx)
+            ray.get([server.set_params.remote(this_shard_ref)])
+
+    def _start_workers(self):
+        """Create worker(actor), maybe need worker group to manager these workers.
+           Or, send these workers to strategy to manager?
+
+           set workers or worker group
+           set worker info, record rank, backend, use_num_gpus?
+        """
+        # TODO (Hao): infer the per-replica batch size here...
+
+        # so here we get two set of params that will be passed around:
+        # (1) Those for setting up training logic in training_operator,
+        # including: batchsize, use_tqdm, user defined operator_config.
+        operator_config = self._operator_config.copy()
+        params = dict(
+            training_operator_cls=self.training_operator_cls,
+            operator_config=operator_config)
+        # (2) params for setting up collective group
+        # and the strategy-related things;
+
+        # For now, we do not have many of them though.
+        dist_params_worker = dict(
+            strategy="ps",
+            is_server=False,
+            group_name="default",
+            num_ps=self.num_ps,
+            num_workers=self.num_workers,
+        )
+
+        dist_params_server = dict(
+            strategy="ps",
+            is_server=True,
+            group_name="default",
+            num_ps=self.num_ps,
+            num_workers=self.num_workers,
+        )
+
+        # (3) other arguments that used to init the DataParallelGrup
+        workergroup_init_args = {
+            "params": params,
+            "dist_params": dist_params_worker,
+            "num_cpus_per_actor": self.num_cpus_per_worker,
+            "num_gpus_per_actor": self.num_gpus_per_worker,
+        }
+
+        servergroup_init_args = {
+            "params": params,
+            "dist_params": dist_params_server,
+            "num_cpus_per_actor": self.num_cpus_per_server,
+            "num_gpus_per_actor": self.num_gpus_per_server,
+        }
+
+        # Should we make two groups for worker and server?
+        self.worker_group = DataParallelGroup(**workergroup_init_args)
```

Comment:
This is strange. Is this the same `DataParallelGroup` with the one in AllReduceStrategy?
If yes -- then fine
If not -- is there any way we can share the same class? If it is hard then we should at least use a different class name?